### PR TITLE
Add reconcile storage cluster function

### PR DESCRIPTION
**Pull request by bindrad**
This PR introduces the reconcileStorageCluster function that will create a provider storageCluster.

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -5,76 +5,126 @@ go 1.17
 require (
 	github.com/go-logr/logr v1.2.3
 	github.com/onsi/ginkgo/v2 v2.1.3
-	github.com/onsi/gomega v1.17.0
+	github.com/onsi/gomega v1.19.0
 	github.com/openshift/api v3.9.1-0.20190924102528-32369d4db2ad+incompatible
-	github.com/operator-framework/api v0.10.0
+	github.com/operator-framework/api v0.16.0
```
Comment (ezio-auditore) [inline go.mod:10]: updation of operator framework api should be in a different PR

File: controllers/managedfusionoffering_controller.go [hunk 2]
```diff
@@ -85,13 +97,118 @@ func pluginReconcile(r *ManagedFusionOfferingReconciler) (ctrl.Result, error) {
...
+	storageCluster.Name = "ocs-storagecluster"
+	storageCluster.Namespace = r.managedFusionOffering.Namespace
+	_, err := ctrl.CreateOrUpdate(r.ctx, r.Client, storageCluster, func() error {
+		if err := r.own(storageCluster); err != nil {
+			return err
+		}
+		sizeAsString := r.managedFusionOffering.Spec.Config["usableCapacityInTiB"]
+
+		// Setting hardcoded value here to force no MCG deployment
+		enableMCGAsString := "false"
```
Comment (ezio-auditore) [inline controllers/managedfusionoffering_controller.go:119]: WE don't need MCG here

File: controllers/managedfusionoffering_controller.go [hunk 3]
```diff
@@ -85,13 +97,118 @@ func pluginReconcile(r *ManagedFusionOfferingReconciler) (ctrl.Result, error) {
...
+	r.Log.Info("Setting storage device set count", "Current", currDeviceSetCount, "New", desiredDeviceSetCount)
+	if currDeviceSetCount <= desiredDeviceSetCount {
+		deviceSet.Count = desiredDeviceSetCount
+	} else {
+		r.Log.V(-1).Info("Requested storage device set count will result in downscaling, which is not supported. Skipping")
+		deviceSet.Count = currDeviceSetCount
+	}
+}
+
+func ensureMCGDeployment(r *ManagedFusionOfferingReconciler, storageCluster *ocsv1.StorageCluster, mcgEnable bool) error {
```
Comment (ezio-auditore) [inline controllers/managedfusionoffering_controller.go:186]: MCG is not required in DF

**Review by ezio-auditore — CHANGES_REQUESTED** [review]

File: controllers/managedfusionoffering_controller.go [hunk 4]
```diff
@@ -85,13 +97,118 @@ func pluginReconcile(r *ManagedFusionOfferingReconciler) (ctrl.Result, error) {
 	if _, found := r.managedFusionOffering.Spec.Config["onboardingValidationKey"]; !found {
 		return ctrl.Result{}, fmt.Errorf("%s CR does not contain onboardingValidationKey entry", r.managedFusionOffering.Name)
 	}
+	if err := reconcileStorageCluster(r); err != nil {
```
Comment (leelavg) [inline controllers/managedfusionoffering_controller.go:100]: - the func signature seems a bit different, is this intentional?

File: controllers/managedfusionoffering_controller.go (hunk 2 above)
Comment (leelavg) [reply controllers/managedfusionoffering_controller.go:119]: curious why we don't need it here 🤔

File: go.mod (hunk 1 above)
Comment (bindrad) [reply go.mod:10]: We are not updating that package as it is mentioned in replaces section
https://github.com/red-hat-storage/managed-fusion-agent/pull/18/files#diff-33ef32bf6c23acb95f5902d7097b7a1d5128ca061167ec0716715b0b9eeaa5f6R124

File: controllers/managedfusionoffering_controller.go (hunk 4 above)
Comment (bindrad) [reply controllers/managedfusionoffering_controller.go:100]: reconcileStorageCluster was a reconciler function in deployer but in agent we are going to have plugin system where each offering can plug into this controller and do whatever is required to install and configure their offering. So we are passing reconciler to these reconcile functions.

File: controllers/managedfusionoffering_controller.go (hunk 2 above)
Comment (bindrad) [reply controllers/managedfusionoffering_controller.go:119]: I think there is no harm in keeping MCG, anyway we are not enabling it.

File: controllers/managedfusionoffering_controller.go (hunk 3 above)
Comment (bindrad) [reply controllers/managedfusionoffering_controller.go:186]: I think there is no harm in keeping MCG, anyway we are not enabling it.

**Review by leelavg — DISMISSED** [review]

File: controllers/managedfusionoffering_controller.go (hunk 3 above)
Comment (ezio-auditore) [reply controllers/managedfusionoffering_controller.go:186]: ack

**Review by ezio-auditore — APPROVED** [review]

**Review by nb-ohad — CHANGES_REQUESTED** [review]

File: controllers/datafoundation.go [hunk 5]
```diff
@@ -21,11 +31,19 @@ type dataFoundationReconciler struct {
...
+//+kubebuilder:rbac:groups=ocs.openshift.io,namespace=system,resources=storageclusters,verbs=get;list;watch;create;update;patch;delete
+
 func dfSetupWatches(controllerBuilder *builder.Builder) {
 	controllerBuilder.
-		Owns(&corev1.Secret{})
+		Owns(&corev1.Secret{}).
+		Owns(&ocsv1.StorageCluster{})
+}
+
+func DFAddToScheme(scheme *runtime.Scheme) {
```
[the remark this answers is not in the document]
Comment (bindrad) [reply controllers/datafoundation.go:45]: `DFAddToScheme` is referenced in another package(main) so it needs to public function whereas other functions are referenced in the same controllers package so they are private.
https://github.com/red-hat-storage/managed-fusion-agent/pull/18/files#diff-2873f79a86c0d8b3335cd7731b0ecf7dd4301eb19a82ef7a1cba7589b5252261R155-R157

File: controllers/datafoundation.go [hunk 6]
```diff
@@ -114,3 +139,61 @@ func (r *dataFoundationReconciler) reconcileOnboardingValidationSecret() error {
...
 }
+
+func (r *dataFoundationReconciler) reconcileStorageCluster() error {
+	r.Log.Info("Reconciling StorageCluster")
+
+	_, err := ctrl.CreateOrUpdate(r.Ctx, r.Client, r.storageCluster, func() error {
+		if err := r.own(r.storageCluster); err != nil {
+			return err
+		}
+		r.Log.Info("Requested add-on settings", usableCapacityInTiBKey, r.spec.usableCapacityInTiB)
```
[the remark this answers is not in the document]
Comment (bindrad) [reply controllers/datafoundation.go]: Removed the log

File: controllers/datafoundation.go [hunk 7]
```diff
@@ -114,3 +139,61 @@ func (r *dataFoundationReconciler) reconcileOnboardingValidationSecret() error {
...
+		desiredDeviceSetCount := int(math.Ceil(float64(r.spec.usableCapacityInTiB) / templates.ProviderOSDSizeInTiB))
+
+		// Get the storage device set count of the current storage cluster
+		currDeviceSetCount := 0
+		if desiredStorageDeviceSet := findStorageDeviceSet(r.storageCluster.Spec.StorageDeviceSets, deviceSetName); desiredStorageDeviceSet != nil {
+			currDeviceSetCount = desiredStorageDeviceSet.Count
+		}
+
+		// Get the desired storage device set from storage cluster template
+		sc := templates.ProviderStorageClusterTemplate.DeepCopy()
```
[the remark this answers is not in the document]
Comment (bindrad) [reply controllers/datafoundation.go]: Created a package for DF offering and keeping storageCluster template in that package

File: controllers/datafoundation.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (nb-ohad) [reply controllers/datafoundation.go:44]: ok

File: offerings/datafoundation/storagecluster.go
Comment (nb-ohad) [inline offerings/datafoundation/storagecluster.go:1]: Why do we need `offerings` in the package name? There might be multiple files here, so `storagecluster.go` is a bad name to take  over, unless you also have a `templates` package.  My suggestion is to name the file `templates_storagecluster` or do `../templates/sotragecluster.go`

**Review by nb-ohad — CHANGES_REQUESTED** [review]

File: offerings/datafoundation/storagecluster.go
Comment (bindrad) [reply offerings/datafoundation/storagecluster.go:1]: Which package has the term `offerings`? I created a folder called offerings and inside that, there is a package called datafoundation for the DF offering.
Why a separate offerings folder? --> Because I don't think it is good idea to have a folder for each and every offering at the top level of the repo.

>  My suggestion is to name the file templates_storagecluster or do ../templates/sotragecluster.go

I will make this change

File: datafoundation/templates_storagecluster.go [hunk 8]
```diff
@@ -0,0 +1,168 @@
...
+			},
+			PreparePlacement: rook.Placement{
+				TopologySpreadConstraints: []corev1.TopologySpreadConstraint{
+					commonTSC,
+					preparePlacementTSC,
+				},
+			},
+			Portable:    true,
+			Replica:     3,
+			DeviceClass: "ssd",
```
Comment (nb-ohad) [inline datafoundation/templates_storagecluster.go:129]: Why is this called ssd? was this the same way with the deployer?
Also, can we move this and `replica` after `count`?

File: controllers/datafoundation.go [hunk 9]
```diff
@@ -2,15 +2,24 @@ package controllers
...
+	ocsv1 "github.com/red-hat-storage/ocs-operator/api/v1"
 	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
 	ctrl "sigs.k8s.io/controller-runtime"
 	"sigs.k8s.io/controller-runtime/pkg/builder"
 )
 
+const (
+	deviceSetName = "default"
```
Comment (nb-ohad) [inline controllers/datafoundation.go:20]: Can we give this a better name? what about `defaultDeviceSetName`? 
To convey meaning and not just what it is

File: datafoundation/templates_storagecluster.go (hunk 8 above)
Comment (bindrad) [reply datafoundation/templates_storagecluster.go:129]: We didn't have DeviceClass in the deployer. Migration team gave me this config.
Yes, I can move the fields but cannot rename the DeviceClass

File: controllers/datafoundation.go (hunk 9 above)
Comment (bindrad) [reply controllers/datafoundation.go]: Updated

**Review by nb-ohad — DISMISSED** [review]

**Review by nb-ohad — APPROVED** [review]
